### PR TITLE
Introduce predicated iteration functions

### PR DESCRIPTION
**Pull request by Ri0n**
Reading a Matio variable from an HDF5 files sometimes can be too slow. Besides it's often necessary to iterate over all variable to read just one. This takes a lot of time. This change allows to skip unnecessary readings

### CODE REVIEW

**Comment by tbeu** [comment]
@Ri0n Can you please force-push again to trigger the Travis CI pipeline. Thanks.

File: src/matio.h
Comment (tbeu) [inline src/matio.h:338]: Is it really necessary to change the public API of matio if only relevent for MAT 7.3?

File: src/mat73.c
Comment (tbeu) [inline src/mat73.c:3186]: Please use a C89 style comment.

```suggestion
    if (mat_data && mat_data->pred && mat_data->pred(name, mat_data->pred_user_data) == 0)  /* do we need to skip it? */
```

File: src/mat.c
Comment (tbeu) [inline src/mat.c:2797]: Please use a C89 style comment.

```suggestion
    } while (pred && pred(matvar->name, user_data) == 0);  /* for 7.3 the predicate will be called one extra time */
```

**Review by tbeu — COMMENTED** [review]
Thanks for the contribution! Do you have any benchmarks demonstrating the performance gain?

File: src/matio.h
Comment (Ri0n) [reply src/matio.h:338]: `Mat_VarReadNextPredicate` was required in a project where we have matio integrated.
Not the case for `Mat_VarReadNextInfoPredicate` though. I can hide it.

Comment (tbeu) [reply src/matio.h:338]: Should be symmetric. So I am OK to keep as is.

**Comment by tbeu** [comment]
If you want to you can add yourself as contributor to section 1.2 of both files README.md and README.

**Comment by Ri0n** [comment]
Wrt performance.
There was a problem with one file. And this PR is an alternative fix to one merged in  https://github.com/tbeu/matio/pull/198
In both cases file loading time was reduced from 2 minutes to 5-7 seconds.

The crux of the problem is in long string string members in `header` variable. 
IIRC 5+ strings. The strings themselves were quite short but declared at 1000+ symbols.
Each symbol was resolved as a variable with `H5Iget_name` somewhere on the way, making all the reading extremely slow.

One way to solve this was to read `header` directly with HDF5 and use this PR's fix to read everything else but `header` with help of matio.  And basically it's what currently applied in our project.

**Comment by tbeu** [comment]
Thanks for the numbers and update.

Do you suggest instead to not merge this PR after #198 got merged?

**Comment by Ri0n** [comment]
I think both together are good to raise performance to its possible maximum.

**Comment by tbeu** [comment]
Argh, I missed that #198 failed the test suite in 18 tests, see https://app.travis-ci.com/github/tbeu/matio/jobs/593339008#L5197

Unfortunately, this is hidden and not marked as failure by Travis CI.

**Comment by Ri0n** [comment]
> Argh, I missed that https://github.com/tbeu/matio/pull/198 failed the test suite in 18 tests, see https://app.travis-ci.com/github/tbeu/matio/jobs/593339008#L5197

That what I was afraid of. It's touches too many places.

**Comment by tbeu** [comment]
Never mind. Do you think you are able to fix the tests?

**Comment by Ri0n** [comment]
Probably I can fix it with some guidance. Unfortunately it's no more a priority task at my job since an alternate fix exists. So I need to invest my own time which I'm always lack of.
In any case I can do nothing till the end of the week.

**Review by tbeu — CHANGES_REQUESTED** [review]
Testsuite is failing by 18 tests (same as #198).

**Comment by tbeu** [comment]
CI succeeds now.

**Pull request merged**
